packages/teams-js/src/public/app.ts

```diff
@@ -806,6 +816,100 @@ export namespace app {
       resolve(sendAndHandleStatusAndReason('executeDeepLink', deepLink));
     });
   }
+
+  /**
+   * Namespace to opt into full-caching or delayed-termination of an app.
+   *
+   * @beta
+   */
+  export namespace lifecycle {
+    /**
+     * Register on resume handler function type
+     *
+     * @param context - Data structure to be used to pass the context to the app.
+     */
+    type registerOnResumeHandlerFunctionType = (context: ResumeContext) => void;
+    /**
+     * A function that is called when the app is ready to be suspended or terminated.
+     */
+    type ReadyToSuspendOrTerminateFunction = () => void;
+
+    /**
+     * Register before suspendOrTerminate handler function type
+     *
+     * @param readyToSuspendOrTerminate - A function that the handler is called with. Apps should call this function when they are ready.
+     * @returns void
+     */
+    type registerBeforeSuspendOrTerminateHandlerFunctionType = (
+      readyToSuspendOrTerminate: ReadyToSuspendOrTerminateFunction,
+    ) => void;
+
+    /**
+     * Registers a handler to be called before the page is suspended or terminatd.
+     *
+     * @param handler - The handler to invoke before the page is suspended or terminated. The page should
+     * invoke the readyToSuspendOrTerminate function provided to it once it's ready.
+     *
+     */
+    export function registerBeforeSuspendOrTerminateHandler(
+      handler: registerBeforeSuspendOrTerminateHandlerFunctionType,
```

Comment:
We can abstractify the function and not expose the `readyToSuspendOrTerminate` at all. 
Once the suspendOrTerminate event is dispatched from host sdk, teams-js would trigger the `registerBeforeSuspendOrTerminateHandler` handler. Right next, teams-js will send `readyToSuspendOrTerminate` back to host. Since the `registerBeforeSuspendOrTerminateHandler` handler is synchronous,  `readyToSuspendOrTerminate` won't be sent until handler execution is done. This way, we can keep the registerHandler as of empty function type, so app developers won't need to deal with all the complexity.
`type registerBeforeSuspendOrTerminateHandlerFunctionType = () => void;`
 @erinha @alexneyman-MSFT